Assets/Mirage/Runtime/NetworkSceneManager.cs

```diff
@@ -182,210 +173,468 @@ internal void OnClientChangeScene(string scenePath, SceneOperation sceneOperatio
+        ///     any other scenes that are loaded on server.
+        ///
+        ///     <para>Default implementation takes main activate scene as main and any other loaded scenes and sends it to player's
+        ///     Please override this function if this is not intended behavior for you.</para>
+        /// </summary>
+        /// <param name="player">The current player that finished authenticating.</param>
+        public virtual void OnServerAuthenticated(INetworkPlayer player)
+        {
+            logger.Log("[NetworkSceneManager] - OnServerAuthenticated");
+
+            string[] scenesPaths = new string[SceneManager.sceneCount - 1];
+
+            for (int sceneIndex = 0; sceneIndex < SceneManager.sceneCount; sceneIndex++)
+            {
+                if (SceneManager.GetActiveScene().buildIndex == sceneIndex) continue;
+
+                Scene scene = SceneManager.GetSceneAt(sceneIndex);
+
+                if (scene.name.Equals(SceneManager.GetActiveScene().name)) continue;
+
+                scenesPaths[sceneIndex == 0 ? sceneIndex : sceneIndex - 1] = scene.path;
+            }
 
-            ServerChangeScene?.Invoke(scenePath, operation);
+            player.Send(new SceneMessage { MainActivateScene = ActiveScenePath, AdditiveScenes = scenesPaths });
+            player.Send(new SceneReadyMessage());
         }
 
         /// <summary>
-        /// Called on the server when a scene is completed loaded, when the scene load was initiated by the server with ChangeServerScene().
+        ///     Called on the server when a scene is completed loaded, when the scene load was initiated by the server with ChangeServerScene().
         /// </summary>
         /// <param name="scenePath">The name of the new scene.</param>
-        internal void OnServerSceneChanged(string scenePath, SceneOperation operation)
+        /// <param name="operation">The type of scene loading we want to do.</param>
+        internal void OnServerFinishedSceneLoad(string scenePath, SceneOperation operation)
         {
-            logger.Log("OnServerSceneChanged");
+            logger.Log(" [NetworkSceneManager] - OnServerSceneChanged");
 
             Server.SendToAll(new SceneReadyMessage());
 
-            ServerSceneChanged?.Invoke(scenePath, operation);
+            ServerFinishedSceneChange?.Invoke(scenePath, operation);
+        }
+
+        /// <summary>
+        ///     When player disconnects from server we will need to clean up info in scenes related to user.
+        /// <para>Default implementation we loop through list of scenes and find where this player was in and removed them from list.</para>
+        /// </summary>
+        /// <param name="disconnectedPlayer"></param>
+        protected virtual void OnServerPlayerDisconnected(INetworkPlayer disconnectedPlayer)
+        {
+            foreach (KeyValuePair<Scene, List<INetworkPlayer>> scene in ServerSceneData)
```

Comment:
maybe the ServerSceneData should use hashsets instead of list so that is is easier to remove players

Reply:
done

Reply:
done